app/models/user.rb

```diff
@@ -39,4 +38,13 @@ def combine_roles_from_tokens
     roles = self.jwt_api_entreprise.reduce([]) { |result, jwt| result + jwt.access_roles }
     roles.uniq
   end
+
+  def random_token_for(attr)
+    constraint = {}
+    loop do
+      token = SecureRandom.hex(10)
+      constraint[attr] = token
+      return token unless User.find_by(constraint)
+    end
```

Comment:
Je ne suis pas certain d'avoir compris l'intérêt de cette boucle.

C'est au cas où on aurait déjà généré la même chaîne ? Ça vaudrait bien un commentaire ou même un test (ça n'a pas l'air simple à écrire par contre...)

Reply:
L'objectif c'est de générer un token aléatoirement, et la boucle c'est simplement pour éviter d'avoir déjà le même token en base attaché à un autre utilisateur : ce serait gênant vu qu'on s'en sert d'identifiant pour identifier l'utilisateur qui demande son renouvellement de mot de passe.

Fondamentalement il y a quand même peu de chances qu'il y ait un jour une collision, mais comme c'est critique la boucle rajoute cette sécurité sur l'unicité en base du jeton.

Concernant le test je ne pense pas qu'il soit nécessaire... Le seul risque c'est qu'on ne sorte jamais de la boucle, et c'est implicitement testé que ça n'arrive pas quand on test qu'un jeton est set en base lors de la demande de renouvellement de mot de passe.

Mais si tu y tiens ce n'est pas si dur que ça à écrire : 

```ruby 
User.take.update(pwd_renewal_token: 'existing_token') 
allow(SecureRandom).to receive(:hex).and_return('existing_token', 'non_existing_token') 
subject 

expect(tested_user.pwd_renewal_token).to eq('non_existing_token')
```

Cf [doc](https://relishapp.com/rspec/rspec-mocks/v/3-3/docs/configuring-responses/returning-a-value#specify-different-return-values-for-multiple-calls)